packages/vm/src/config.rs

```diff
@@ -0,0 +1,150 @@
+        Self {
+            wasm_limits: WasmLimits::default(),
+            cache,
+        }
+    }
+}
+
+/// Limits for static validation of Wasm files. These are checked before storing the Wasm file.
+/// All limits are optional because they are coming from the Go-side and have default values.
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+#[non_exhaustive]
+pub struct WasmLimits {
+    /// Maximum number of memory pages that a module can request.
+    pub memory_page_limit: Option<u32>,
+    /// The upper limit for the `max` value of each table. CosmWasm contracts have
+    /// initial=max for 1 table. See
+    ///
+    /// ```plain
+    /// $ wasm-objdump --section=table -x packages/vm/testdata/hackatom.wasm
+    /// Section Details:
+    ///
+    /// Table[1]:
+    /// - table[0] type=funcref initial=161 max=161
+    /// ```
+    ///
+    pub table_size_limit: Option<u32>,
+    /// If the contract has more than this amount of imports, it will be rejected
+    /// during static validation before even looking into the imports.
+    pub max_imports: Option<usize>,
+
+    /// The maximum number of functions a contract can have.
+    /// Any contract with more functions than this will be rejected during static validation.
+    pub max_functions: Option<usize>,
+
+    /// The maximum number of parameters a Wasm function can have.
+    pub max_function_params: Option<usize>,
+    /// The maximum total number of parameters of all functions in the Wasm.
+    /// For each function in the Wasm, take the number of parameters and sum all of these up.
+    /// If that sum exceeds this limit, the Wasm will be rejected during static validation.
+    ///
+    /// Be careful when adjusting this limit, as it prevents an attack where a small Wasm file
+    /// explodes in size when compiled.
+    pub max_total_function_params: Option<usize>,
+
+    /// The maximum number of results a Wasm function type can have.
+    pub max_function_results: Option<usize>,
+}
+
+impl WasmLimits {
+    pub fn memory_page_limit(&self) -> u32 {
+        self.memory_page_limit.unwrap_or(DEFAULT_MEMORY_LIMIT)
+    }
+
+    pub fn table_size_limit(&self) -> u32 {
+        self.table_size_limit.unwrap_or(DEFAULT_TABLE_SIZE_LIMIT)
+    }
+
+    pub fn max_imports(&self) -> usize {
+        self.max_imports.unwrap_or(DEFAULT_MAX_IMPORTS)
+    }
+
+    pub fn max_functions(&self) -> usize {
+        self.max_functions.unwrap_or(DEFAULT_MAX_FUNCTIONS)
+    }
+
+    pub fn max_function_params(&self) -> usize {
+        self.max_function_params
+            .unwrap_or(DEFAULT_MAX_FUNCTION_PARAMS)
+    }
+
+    pub fn max_total_function_params(&self) -> usize {
+        self.max_total_function_params
+            .unwrap_or(DEFAULT_MAX_TOTAL_FUNCTION_PARAMS)
+    }
+
+    pub fn max_function_results(&self) -> usize {
+        self.max_function_results
+            .unwrap_or(DEFAULT_MAX_FUNCTION_RESULTS)
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
```

Comment:
Is this added Serialize, Deserialize needed somewhere?

Reply:
It's needed in wasmvm where the msgpack is deserialized using this.

Reply:
But CacheOptions have always been created in Rust from the individual arguments. Are you planning to change that?

Reply:
Yes, at least the idea was to pass the whole config using msgpack.